### PR TITLE
Deprecate Task.deleteAllActions()

### PR DESCRIPTION
**Pull request by lptr**
Removing task actions is a bad idea, we shouldn't support it. This is a step towards making it impossible to remove actions already added to a task.

### CODE REVIEW

**Comment by lptr** [comment]
Feedback build: https://builds.gradle.org/viewQueued.html?itemId=11485829

File: subprojects/core-api/src/main/java/org/gradle/api/Task.java [hunk 1]
```diff
@@ -421,7 +421,10 @@ public String determineName(Task c) {
      * <p>Removes all the actions of this task.</p>
      *
      * @return the task object this method is applied to
+     *
+     * @deprecated Don't use this.
```
Comment (wolfs) [inline subprojects/core-api/src/main/java/org/gradle/api/Task.java]: Maybe we could come up with a better deprecation reason.

File: subprojects/docs/src/docs/release/notes.md [hunk 2]
```diff
@@ -53,6 +53,8 @@ The following are the newly deprecated items in this Gradle release. If you have
 ### Example deprecation
 -->
 
+* `Task.deleteAllActions()` is deprecated without replacement.
```
Comment (wolfs) [inline subprojects/docs/src/docs/release/notes.md]: Do we need a header here or is a list like you started sufficient?

File: subprojects/core-api/src/main/java/org/gradle/api/Task.java (hunk 1 above)
Comment (lptr) [reply subprojects/core-api/src/main/java/org/gradle/api/Task.java]: I don't think we need to sugar-coat this: it is not a useful method, we should just remove it. That said, I'm not against anything if you have a better suggestion.

File: subprojects/docs/src/docs/release/notes.md (hunk 2 above)
Comment (lptr) [reply subprojects/docs/src/docs/release/notes.md]: We could add something like "Other minor deprecations" but if this is the only one in the end, then we will need to rewrite it anyway. I'd leave it like this until before the release when we release notes anyway. The important thing is to capture the deprecation now.

**Comment by lacasseio** [comment]
@lptr Could you expand on why it's a bad idea to remove task actions?

I toyed around with the idea of allowing remote build by simply configuring key tasks to replace its task action with action that would outsource the execution to another machine (aka remote compilation). I'm aware that it probably breaks caching and have a lot more gotcha but `Task.deleteAllActions()` was my escape hatch to allow such experimentation. My point is deprecating `Task.deleteAllActions()` too strong at this stage? Maybe more documentation around why it's bad in most context could be smaller step until Gradle support remote execution natively?

**Comment by lptr** [comment]
There are multiple issues with allowing actions to be deleted, but the main one is that your code deleting actions has to make a lot of assumptions about the implementation of the actions it deletes. Since we wrap action objects in internal wrappers, you can't identify even the actions you added, not to mention other actions added by some third-party plugin or build script code. Calling `deleteAllActions()` is eve worse than trying to delete individual actions, as it means you indiscriminately remove everything. The resulting code is just way too fragile.

**Comment by lacasseio** [comment]
It makes sense, thanks for explaining.

**Pull request merged**